packages/frontend/src/components/CustomVisualization/index.tsx

```diff
@@ -1,107 +1,42 @@
-
-const CustomVisualizationContext = createContext<{
-    chartConfig: string;
-    setChartConfig: (newConfig: string) => void;
-    rows: {
-        [k: string]: unknown;
-    }[];
-    fields: string[];
-}>({
-    chartConfig: defaultValue,
-    setChartConfig: () => {},
-    rows: [],
-    fields: [],
-});
-
-export const useCustomVisualizationContext = () =>
-    useContext(CustomVisualizationContext);
-
-const convertRowsToSeries = (rows: ResultRow[]) => {
-    return rows.map((row) => {
-        return Object.fromEntries(
-            Object.entries(row).map(([key, rowValue]) => [
-                key,
-                rowValue.value.raw,
-            ]),
-        );
-    });
-};
-
-export const CustomVisualizationProvider: FC<React.PropsWithChildren<{}>> = ({
-    children,
-}) => {
-    const rows = useExplorerContext(
-        (context) => context.queryResults.data?.rows,
-    );
-
-    const [chartConfig, setChartConfig] = useState<string>(defaultValue);
-
-    const convertedRows = useMemo(() => {
-        return rows ? convertRowsToSeries(rows) : [];
-    }, [rows]);
-
-    const fields = useMemo(() => {
-        return rows && rows.length > 0 ? Object.keys(rows[0]) : [];
-    }, [rows]);
-
-    return (
-        <CustomVisualizationContext.Provider
-            value={{
-                chartConfig,
-                setChartConfig,
-                rows: convertedRows,
-                fields,
-            }}
-        >
-            {children}
-        </CustomVisualizationContext.Provider>
-    );
-};
-
-type CustomVisualizationProps = {
+type Props = {
     className?: string;
     'data-testid'?: string;
 };
 
-const CustomVisualization: FC<CustomVisualizationProps> = (props) => {
-    const { chartConfig, rows } = useCustomVisualizationContext();
+const CustomVisualization: FC<Props> = (props) => {
+    const { isLoading, visualizationConfig } = useVisualizationContext();
+
+    if (isLoading) {
+        return <Text>Loading...</Text>;
+    }
 
-    const [config, error] = useMemo(() => {
-        try {
-            return [
-                {
-                    ...JSON.parse(chartConfig),
-                },
-                null,
-            ];
-        } catch (e) {
-            return [null, e];
-        }
-    }, [chartConfig]);
+    if (!isCustomVisualizationConfig(visualizationConfig)) return null;
+    const spec = visualizationConfig.chartConfig.validConfig.spec;
 
-    if (error) {
-        return <Code>{error.toString()}</Code>;
+    if (
+        !visualizationConfig ||
+        !isCustomVisualizationConfig(visualizationConfig) ||
+        !spec
+    ) {
+        return null;
     }
 
-    const data = { table: rows };
+    // TODO: 'chartConfig' is more props than config. It has data and
+    // configuration for the chart. We should consider renaming it generally.
+    const visProps =
+        visualizationConfig.chartConfig as CustomVisualizationConfigAndData;
```

Comment:
you don't need `as CustomVisualizationConfigAndData` because `if` condition above with `isCustomVisualizationConfig(visualizationConfig)` already narrows down the type for you